zenoh/src/api/builders/close.rs

```diff
@@ -0,0 +1,151 @@
+        self,
+    ) -> BackgroundCloseBuilder<<CloseBuilder<TCloseable> as Resolvable>::To> {
+        BackgroundCloseBuilder::new(self.into_future())
+    }
+}
+
+impl<TCloseable: Closeable> Resolvable for CloseBuilder<TCloseable> {
+    type To = ZResult<()>;
+}
+
+impl<TCloseable: Closeable> Wait for CloseBuilder<TCloseable> {
+    fn wait(self) -> Self::To {
+        ZRuntime::Application.block_in_place(self.into_future())
+    }
+}
+
+impl<TCloseable: Closeable> IntoFuture for CloseBuilder<TCloseable> {
+    type Output = <Self as Resolvable>::To;
+    type IntoFuture = Pin<Box<dyn Future<Output = <Self as IntoFuture>::Output> + Send>>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(
+            async move {
+                if tokio::time::timeout(self.timeout, self.closee.close_inner())
+                    .await
+                    .is_err()
+                {
+                    bail!("close operation timed out!")
+                }
+                Ok(())
+            }
+            .into_future(),
+        )
+    }
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+/// A builder for close operations running in background
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[zenoh_macros::unstable_doc]
+#[allow(private_bounds)]
+pub struct BackgroundCloseBuilder<TOutput: Send + 'static> {
+    inner: Pin<Box<dyn Future<Output = TOutput> + Send>>,
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+// NOTE: `Closeable` is only pub(crate) because it is zenoh-internal trait, so we don't
+// care about the `private_bounds` lint in this particular case.
+#[allow(private_bounds)]
+impl<TOutput: Send + 'static> BackgroundCloseBuilder<TOutput> {
+    fn new(inner: Pin<Box<dyn Future<Output = TOutput> + Send>>) -> Self {
+        Self { inner }
+    }
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+impl<TOutput: Send + 'static> Resolvable for BackgroundCloseBuilder<TOutput> {
+    type To = tokio::task::JoinHandle<TOutput>;
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+impl<TOutput: Send + 'static> Wait for BackgroundCloseBuilder<TOutput> {
+    fn wait(self) -> Self::To {
+        ZRuntime::Application.block_in_place(self.into_future())
+    }
+}
+
+#[cfg(all(feature = "unstable", feature = "internal"))]
+impl<TOutput: Send + 'static> IntoFuture for BackgroundCloseBuilder<TOutput> {
+    type Output = <Self as Resolvable>::To;
+    type IntoFuture = Pin<Box<dyn Future<Output = <Self as IntoFuture>::Output> + Send>>;
+
+    // NOTE: yes, we need to return a future that returns JoinHandle
+    #[allow(clippy::async_yields_async)]
+    fn into_future(self) -> Self::IntoFuture {
+        Box::pin(async move { ZRuntime::Application.spawn(self.inner) }.into_future())
+    }
+}
+
+#[async_trait]
+pub(crate) trait Closee: Send + Sync + 'static {
```

Comment:
Very weird name. What does `Closee` mean?

Reply:
> Very weird name. What does `Closee` mean?

https://dictionary.cambridge.org/dictionary/english/ee

The quick search showed there is no 'closee' word, however the '-ee' added to 'close' is smth that perfectly reflects the trait's purpose.

Reply:
Ok, still it's not clear what the trait implementer is expected to do and who should implement it.

Reply:
Mmmm not convinced either. Anyway, I don't think it's a blocker. LGTM.